File: java/testsuite/ApplicationCases/SceneEditorApplication/src/model/ApplicationModel.java
package model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import vsdk.toolkit.gui.viewport.ViewportSet;
import vsdk.toolkit.gui.widget.Widget;
import vsdk.toolkit.environment.camera.Camera;
import vsdk.toolkit.environment.geometry.element.Ray;
import vsdk.toolkit.environment.light.Light;
import vsdk.toolkit.environment.light.PointLight;
import vsdk.toolkit.environment.scene.SimpleBody;
import vsdk.toolkit.media.RGBColorPalette;
import vsdk.toolkit.media.RGBImageUncompressed;
import vsdk.toolkit.media.ZBuffer;

public class ApplicationModel
{
    private Scene scene;
    private RGBImageUncompressed raytracedImage;
    /// OpenGL depth of `raytracedImage` when it is shown in a viewport
    private ZBuffer raytracedDepth;
    private RGBImageUncompressed zbufferImage;
    private int raytracedImageWidth;
    private int raytracedImageHeight;
    private RGBColorPalette palette;
    private boolean withVisualDebugRay;
    private Ray visualDebugRay;
    private int visualDebugRayLevels;
    private final List<ViewportSet> viewportSets;
    private int activeViewportSetIndex;
    private Widget i18nContext;
    private final SceneLightFactory lightFactory = new SceneLightFactory();
    private final DrawingArea drawingArea;
    private final GuiState guiState = new GuiState();

    /**
    Creates the model with one standard `ViewportSet`. More sets can be added
    to support users working with several displays.
    */
    public ApplicationModel()
    {
        viewportSets = new ArrayList<ViewportSet>();
        viewportSets.add(ViewportSet.createStandardSet("Display 1"));
        activeViewportSetIndex = 0;
        drawingArea = new DrawingArea(viewportSets.get(0));
    }

    /**
    @return state of the GUI that does not depend on the GUI technology
    */
    public GuiState getGuiState()
    {
        return guiState;
    }

    /**
    @return the drawing area presenting the viewport set of the first display
    */
    public DrawingArea getDrawingArea()
    {
        return drawingArea;
    }

    /**
    @return a read-only view of the viewport sets, one for each display
    */
    public List<ViewportSet> getViewportSets()
    {
        return Collections.unmodifiableList(viewportSets);
    }

    public void addViewportSet(ViewportSet viewportSet)
    {
        if ( viewportSet != null ) {
            viewportSet.setI18nContext(i18nContext);
            viewportSets.add(viewportSet);
        }
    }

    /**
    @return the I18N context (GUI definition in the current language) shared by
    the viewport sets, or null if there is none yet
    */
    public Widget getI18nContext()
    {
        return i18nContext;
    }

    /**
    Sets the I18N context and propagates it to every viewport set, so they are
    presented with the messages of the language currently selected by the
    user. It must be called each time the GUI definition is loaded.
    @param i18nContext
    */
    public void setI18nContext(Widget i18nContext)
    {
        this.i18nContext = i18nContext;
        for ( ViewportSet viewportSet : viewportSets ) {
            viewportSet.setI18nContext(i18nContext);
        }
    }

    public int getActiveViewportSetIndex()
    {
        return activeViewportSetIndex;
    }

    public void setActiveViewportSetIndex(int activeViewportSetIndex)
    {
        if ( activeViewportSetIndex >= 0 && activeViewportSetIndex < viewportSets.size() ) {
            this.activeViewportSetIndex = activeViewportSetIndex;
        }
    }

    /**
    @return the viewport set of the display currently receiving interaction
    */
    public ViewportSet getActiveViewportSet()
    {
        return viewportSets.get(activeViewportSetIndex);
    }

    public Scene getScene()
    {
        return scene;
    }

    public void setScene(Scene scene)
    {
        this.scene = scene;
    }

    public Camera getCamera()
    {
        return scene.camera;
    }

    public Camera getActiveCamera()
    {
        return scene.activeCamera;
    }

    public void setActiveCamera(Camera activeCamera)
    {
        scene.activeCamera = activeCamera;
    }

    public List<Light> getLights()
    {
        return scene.scene.getLights();
    }

    /**
    Adds a new point light to the scene, placed inside the view volume of a
    camera of the active viewport set (see `SceneLightFactory`).
    @return the added light, or null if no viewport is visible
    */
    public PointLight addNewLight()
    {
        PointLight light = lightFactory.createLight(getLights(), getActiveViewportSet());

        if ( light != null ) {
            getLights().add(light);
        }
        return light;
    }

    public List<SimpleBody> getSimpleBodies()
    {
        return scene.scene.getSimpleBodies();
    }

    /**
    @return window space (OpenGL) depth of each pixel of the raytraced image
    of a viewport in CPU render mode, or null if there is none
    */
    public ZBuffer getRaytracedDepth()
    {
        return raytracedDepth;
    }

    /**
    @param raytracedDepth window space (OpenGL) depth of each pixel of the
    raytraced image, or null
    */
    public void setRaytracedDepth(ZBuffer raytracedDepth)
    {
        this.raytracedDepth = raytracedDepth;
    }

    public RGBImageUncompressed getRaytracedImage()
    {
        return raytracedImage;
    }

    public void setRaytracedImage(RGBImageUncompressed raytracedImage)
    {
        this.raytracedImage = raytracedImage;
    }

    public RGBImageUncompressed getZbufferImage()
    {
        return zbufferImage;
    }

    public void setZbufferImage(RGBImageUncompressed zbufferImage)
    {
        this.zbufferImage = zbufferImage;
    }

    public int getRaytracedImageWidth()
    {
        return raytracedImageWidth;
    }

    public void setRaytracedImageWidth(int raytracedImageWidth)
    {
        this.raytracedImageWidth = raytracedImageWidth;
    }

    public int getRaytracedImageHeight()
    {
        return raytracedImageHeight;
    }

    public void setRaytracedImageHeight(int raytracedImageHeight)
    {
        this.raytracedImageHeight = raytracedImageHeight;
    }

    public RGBColorPalette getPalette()
    {
        return palette;
    }

    public void setPalette(RGBColorPalette palette)
    {
        this.palette = palette;
    }

    public boolean isWithVisualDebugRay()
    {
        return withVisualDebugRay;
    }

    public void setWithVisualDebugRay(boolean withVisualDebugRay)
    {
        this.withVisualDebugRay = withVisualDebugRay;
    }

    public Ray getVisualDebugRay()
    {
        return visualDebugRay;
    }

    public void setVisualDebugRay(Ray visualDebugRay)
    {
        this.visualDebugRay = visualDebugRay;
    }

    public int getVisualDebugRayLevels()
    {
        return visualDebugRayLevels;
    }

    public void setVisualDebugRayLevels(int visualDebugRayLevels)
    {
        this.visualDebugRayLevels = visualDebugRayLevels;
    }
}

File: cpp/testsuite/ApplicationCases/SceneEditorApplication/src/model/ApplicationModel.cpp
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/environment/light/PointLight.h"
#include "vsdk/toolkit/environment/scene/SimpleBody.h"
#include "vsdk/toolkit/environment/scene/SimpleScene.h"
#include "vsdk/toolkit/gui/viewport/ViewportSet.h"
#include "vsdk/toolkit/gui/widget/Widget.h"
#include "vsdk/toolkit/media/RGBColorPalette.h"
#include "vsdk/toolkit/media/RGBImageUncompressed.h"
#include "vsdk/toolkit/media/ZBuffer.h"
#include "model/ApplicationModel.h"
#include "model/Scene.h"

ApplicationModel::ApplicationModel()
    : scene(nullptr), raytracedImage(nullptr), raytracedDepth(nullptr),
      zbufferImage(nullptr), raytracedImageWidth(0), raytracedImageHeight(0),
      palette(nullptr), withVisualDebugRay(false), hasVisualDebugRay(false),
      visualDebugRayLevels(0), activeViewportSetIndex(0),
      i18nContext(nullptr), drawingArea(nullptr),
      editHistory([this]() { return getScene(); })
{
    viewportSets.add(ViewportSet::createStandardSet("Display 1"));
    activeViewportSetIndex = 0;
    drawingArea = new DrawingArea(viewportSets.get(0));
}

ApplicationModel::~ApplicationModel()
{
    long i;

    // The history may reference the scene and the viewports
    editHistory.getSceneHistory()->clear();
    editHistory.getViewportHistory()->clear();
    delete drawingArea;
    for ( i = 0; i < viewportSets.size(); i++ ) {
        delete viewportSets.get(i);
    }
    delete scene;
    delete i18nContext;
    delete palette;
    delete raytracedImage;
    delete raytracedDepth;
    delete zbufferImage;
}

GuiState* ApplicationModel::getGuiState()
{
    return &guiState;
}

DrawingArea* ApplicationModel::getDrawingArea() const
{
    return drawingArea;
}

const java::ArrayList<ViewportSet*>& ApplicationModel::getViewportSets() const
{
    return viewportSets;
}

void ApplicationModel::addViewportSet(ViewportSet* viewportSet)
{
    if ( viewportSet != nullptr ) {
        viewportSet->setI18nContext(i18nContext);
        viewportSets.add(viewportSet);
    }
}

Widget* ApplicationModel::getI18nContext() const
{
    return i18nContext;
}

void ApplicationModel::setI18nContext(Widget* i18nContext)
{
    if ( this->i18nContext != i18nContext ) {
        delete this->i18nContext;
    }
    this->i18nContext = i18nContext;
    long i;
    for ( i = 0; i < viewportSets.size(); i++ ) {
        viewportSets.get(i)->setI18nContext(i18nContext);
    }
}

int ApplicationModel::getActiveViewportSetIndex() const
{
    return activeViewportSetIndex;
}

void ApplicationModel::setActiveViewportSetIndex(int activeViewportSetIndex)
{
    if ( activeViewportSetIndex >= 0 &&
         activeViewportSetIndex < viewportSets.size() ) {
        this->activeViewportSetIndex = activeViewportSetIndex;
    }
}

ViewportSet* ApplicationModel::getActiveViewportSet() const
{
    return viewportSets.get(activeViewportSetIndex);
}

Scene* ApplicationModel::getScene() const
{
    return scene;
}

void ApplicationModel::setScene(Scene* scene)
{
    // The operations of the history reference the former scene
    editHistory.getSceneHistory()->clear();
    if ( this->scene != scene ) {
        delete this->scene;
    }
    this->scene = scene;
}

EditHistory* ApplicationModel::getEditHistory()
{
    return &editHistory;
}

Camera* ApplicationModel::getCamera() const
{
    return scene->camera;
}

Camera* ApplicationModel::getActiveCamera() const
{
    return scene->activeCamera;
}

void ApplicationModel::setActiveCamera(Camera* activeCamera)
{
    scene->activeCamera = activeCamera;
}

java::ArrayList<Light*>& ApplicationModel::getLights()
{
    return scene->scene->getLights();
}

PointLight* ApplicationModel::addNewLight()
{
    PointLight* light = lightFactory.createLight(getLights(),
                                                 getActiveViewportSet());

    if ( light != nullptr ) {
        getLights().add(light);
    }
    return light;
}

java::ArrayList<SimpleBody*>& ApplicationModel::getSimpleBodies()
{
    return scene->scene->getSimpleBodies();
}

ZBuffer* ApplicationModel::getRaytracedDepth() const
{
    return raytracedDepth;
}

void ApplicationModel::setRaytracedDepth(ZBuffer* raytracedDepth)
{
    if ( this->raytracedDepth != raytracedDepth ) {
        delete this->raytracedDepth;
    }
    this->raytracedDepth = raytracedDepth;
}

RGBImageUncompressed* ApplicationModel::getRaytracedImage() const
{
    return raytracedImage;
}

void ApplicationModel::setRaytracedImage(RGBImageUncompressed* raytracedImage)
{
    if ( this->raytracedImage != raytracedImage ) {
        delete this->raytracedImage;
    }
    this->raytracedImage = raytracedImage;
}

RGBImageUncompressed* ApplicationModel::getZbufferImage() const
{
    return zbufferImage;
}

void ApplicationModel::setZbufferImage(RGBImageUncompressed* zbufferImage)
{
    if ( this->zbufferImage != zbufferImage ) {
        delete this->zbufferImage;
    }
    this->zbufferImage = zbufferImage;
}

int ApplicationModel::getRaytracedImageWidth() const
{
    return raytracedImageWidth;
}

void ApplicationModel::setRaytracedImageWidth(int raytracedImageWidth)
{
    this->raytracedImageWidth = raytracedImageWidth;
}

int ApplicationModel::getRaytracedImageHeight() const
{
    return raytracedImageHeight;
}

void ApplicationModel::setRaytracedImageHeight(int raytracedImageHeight)
{
    this->raytracedImageHeight = raytracedImageHeight;
}

RGBColorPalette* ApplicationModel::getPalette() const
{
    return palette;
}

void ApplicationModel::setPalette(RGBColorPalette* palette)
{
    if ( this->palette != palette ) {
        delete this->palette;
    }
    this->palette = palette;
}

bool ApplicationModel::isWithVisualDebugRay() const
{
    return withVisualDebugRay;
}

void ApplicationModel::setWithVisualDebugRay(bool withVisualDebugRay)
{
    this->withVisualDebugRay = withVisualDebugRay;
}

const Ray* ApplicationModel::getVisualDebugRay() const
{
    return hasVisualDebugRay ? &visualDebugRay : nullptr;
}

void ApplicationModel::setVisualDebugRay(const Ray* visualDebugRay)
{
    hasVisualDebugRay = visualDebugRay != nullptr;
    if ( hasVisualDebugRay ) {
        this->visualDebugRay = *visualDebugRay;
    }
}

int ApplicationModel::getVisualDebugRayLevels() const
{
    return visualDebugRayLevels;
}

void ApplicationModel::setVisualDebugRayLevels(int visualDebugRayLevels)
{
    this->visualDebugRayLevels = visualDebugRayLevels;
}
